packages/graphql-persisted/src/apollo.ts

```diff
@@ -0,0 +1,109 @@
+import {ApolloLink, Operation, NextLink, Observable} from 'apollo-link';
+import {CacheMissBehavior} from './shared';
+
+interface Options {
+  idFromOperation?(operation: Operation): string | undefined | null;
+}
+
+export function createPersistedLink(options: Options = {}) {
+  return new PersistedLink(options);
+}
+
+export class PersistedLink extends ApolloLink {
+  private sendAlwaysIds = new Set<string>();
+
+  constructor(private options: Options) {
+    super();
+  }
+
+  request(operation: Operation, forward?: NextLink) {
+    if (forward == null) {
+      throw new Error('Persisted link can’t be a terminating link.');
+    }
+
+    return new Observable(observer => {
+      const {idFromOperation = defaultIdFromOperation} = this.options;
+      const id = idFromOperation(operation);
+
+      if (typeof id !== 'string' || this.sendAlwaysIds.has(id)) {
+        const subscription = forward(operation).subscribe(observer);
+        return () => subscription.unsubscribe();
+      }
+
+      operation.extensions.persisted = {
+        id,
+      };
+
+      operation.setContext({
+        http: {
+          includeQuery: false,
+          includeExtensions: true,
+        },
+      });
+
+      let subscription: ReturnType<
+        ReturnType<typeof forward>['subscribe']
+      > | null = null;
+
+      subscription = forward(operation).subscribe({
+        next: response => {
+          const errors = response ? response.errors || [] : [];
```

Comment:
```suggestion
          const errors = response && response.errors || [];
```

But I don't mind either way ¯\\_(ツ)_/¯